src/mlpack/methods/edge_boxes/feature_extraction_impl.hpp

```diff
@@ -0,0 +1,960 @@
+          << 0.178325 << 0.071330 << 0.939180;
+
+  //see how to calculate this efficiently. numpy.dot does this.
+  CubeType xyz(InImage.n_rows, InImage.n_cols, rgb2xyz.n_cols);
+
+  for (size_t i = 0; i < InImage.slice(0).n_elem; ++i)
+  {
+    double r = InImage.slice(0)(i);
+    double g = InImage.slice(1)(i);
+    double b = InImage.slice(2)(i);
+    
+    xyz.slice(0)(i) = 0.430574 * r + 0.341550 * g + 0.178325 * b;
+    xyz.slice(1)(i) = 0.222015 * r + 0.706655 * g + 0.071330 * b;
+    xyz.slice(2)(i) = 0.020183 * r + 0.129553 * g + 0.939180 * b;
+  
+  /*
+    xyz.slice(0)(i) = 0.430574 * r + 0.341550 * g + 0.178325 * b;
+    xyz.slice(1)(i) = 0.222015 * r + 0.706655 * g + 0.129553 * b;
+    xyz.slice(2)(i) = 0.020183 * r + 0.071330 * g + 0.939180 * b;
+  */
+  }
+
+  MatType nz(InImage.n_rows, InImage.n_cols);
+
+  nz = 1.0 / ( xyz.slice(0) + (15 * xyz.slice(1) ) + 
+       (3 * xyz.slice(2) + 1e-35));
+  CubeType OutImage(InImage.n_rows, InImage.n_cols, InImage.n_slices);
+
+  for(size_t j = 0; j < xyz.n_cols; ++j)
+  {
+    for(size_t i = 0; i < xyz.n_rows; ++i)
+    {
+      OutImage(i, j, 0) = table( static_cast<size_t>( (1024 * xyz(i, j, 1) ) ) );
+    }
+  }
+
+  OutImage.slice(1) = OutImage.slice(0) % (13 * 4 * (xyz.slice(0) % nz) \
+              - 13 * 0.197833) + 88 * maxi;
+  OutImage.slice(2) = OutImage.slice(0) % (13 * 9 * (xyz.slice(1) % nz) \
+              - 13 * 0.468331) + 134 * maxi;
+
+  return OutImage; 
+}
+
+/*implement this function in a column major order.*/
+template<typename MatType, typename CubeType>
+MatType StructuredForests<MatType, CubeType>::
+bilinearInterpolation(MatType const &src,
+                      size_t height, size_t width)
+{
+  MatType dst(height, width);
+  double const x_ratio = static_cast<double>((src.n_cols - 1)) / width;
+  double const y_ratio = static_cast<double>((src.n_rows - 1)) / height;
+  for(size_t row = 0; row != dst.n_rows; ++row)
+  {
+    size_t y = static_cast<size_t>(row * y_ratio);
+    double const y_diff = (row * y_ratio) - y; //distance of the nearest pixel(y axis)
+    double const y_diff_2 = 1 - y_diff;
+    for(size_t col = 0; col != dst.n_cols; ++col)
+    {
+      size_t x = static_cast<size_t>(col * x_ratio);
+      double const x_diff = (col * x_ratio) - x; //distance of the nearet pixel(x axis)
+      double const x_diff_2 = 1 - x_diff;
+      double const y2_cross_x2 = y_diff_2 * x_diff_2;
+      double const y2_cross_x = y_diff_2 * x_diff;
+      double const y_cross_x2 = y_diff * x_diff_2;
+      double const y_cross_x = y_diff * x_diff;
+      dst(row, col) = y2_cross_x2 * src(y, x) +
+              y2_cross_x * src(y, x + 1) +
+              y_cross_x2 * src(y + 1, x) +
+              y_cross_x * src(y + 1, x + 1);
+    }
+  }
+
+  return dst;
+}
+
+template<typename MatType, typename CubeType>
+CubeType StructuredForests<MatType, CubeType>::
+sepFilter2D(CubeType &InOutImage, arma::vec &kernel, size_t radius)
+{
```

Comment:
Do you think it's reasonable to use the NaiveConvolution function from the mlpack code?

``NaiveConvolution::Convolution(input, kernel, output)``